File: 0958-check-completeness-of-a-binary-tree/0958-check-completeness-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int countNodes(TreeNode* root){
        if (!root) return 0;
        
        return 1 + countNodes(root -> left) + countNodes(root -> right);
    }
    
    bool dfs(TreeNode* root, int index, int n){
        if (!root) return true;
        if (index >= n) return false;
        
        return dfs(root -> left, 2 * index + 1, n) && dfs(root -> right, 2 * index + 2, n);
    }
    
public:
    bool isCompleteTree(TreeNode* root) {
        return dfs(root, 0, countNodes(root));
    }
};

File: 0958-check-completeness-of-a-binary-tree/0958-check-completeness-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)
    
    def dfs(self, node: Optional[TreeNode], index: int, n: int) -> bool:
        if not node:
            return True
        
        if index >= n:
            return False
        
        return self.dfs(node.left, 2 * index + 1, n) and self.dfs(node.right, 2 * index + 2, n)
    
    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root, 0, self.countNodes(root))